# **AI And Biotechnology/Bioinformatics**

## **AI and Drug Discovery Course: QSAR Modeling**
## **Assignment 2: QSAR Data Curation**

**Student Name:** [YOUR NAME HERE]

This notebook demonstrates how to collect and preprocess bioactivity data from ChEMBL for QSAR modeling.

**Selected Target:** Acetylcholinesterase (AChE) - Alzheimer's Disease


# **Part 1: Data Collection & Curation**


**First we need to connect Google Colab with our Google Drive, so that we can have access to our Google drive within Colab.**

This allows us to:
* Save datasets
* Reload data across sessions
* Organize project files


In [1]:
# =============================================================================
# CELL 4 - CODE: Mount Google Drive
# =============================================================================
from google.colab import drive
drive.mount('/content/gdrive/', force_remount=True)


Mounted at /content/gdrive/


**Now create "data" folder in our "Colab Notebooks" folder on Google Drive.**


In [2]:
# =============================================================================
# CELL 6 - CODE: Create data folder
# =============================================================================
! mkdir "/content/gdrive/My Drive/Colab Notebooks/data"


## Install and Import Required Libraries
We install the ChEMBL web service package so that we can retrieve bioactivity data


In [3]:
# =============================================================================
# CELL 8 - CODE: Install ChEMBL client
# =============================================================================
!pip install chembl_webresource_client


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 55.2/55.2 kB 2.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.4/61.4 kB 4.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 70.7/70.7 kB 4.1 MB/s eta 0:00:00


# Import Libraries
* pandas for data handling
* new_client from chembl for accessing the database


In [4]:
# =============================================================================
# CELL 10 - CODE: Import libraries
# =============================================================================
import pandas as pd
from chembl_webresource_client.new_client import new_client


# Step 1: Search for Target Protein

## **Target Identification (Acetylcholinesterase)**
Search ChEMBL for the Acetylcholinesterase target and select the most relevant entry.

**Why Acetylcholinesterase?**
- Key target for Alzheimer's disease treatment
- Well-characterized enzyme with abundant bioactivity data
- Inhibitors like Donepezil are approved drugs


In [5]:
# =============================================================================
# CELL 12 - CODE: Search for target
# =============================================================================
target = new_client.target
target_query = target.search("acetylcholinesterase")
targets = pd.DataFrame.from_dict(target_query)
targets.head(10)


,cross_references,organism,pref_name,score,species_group_flag,target_chembl_id,target_components,target_type,tax_id
0,[],Drosophila melanogaster,Acetylcholinesterase,18.0,False,CHEMBL2242744,"[{'accession': 'P07140', 'component_descriptio...",SINGLE PROTEIN,7227
1,[],Homo sapiens,Acetylcholinesterase,16.0,False,CHEMBL220,"[{'accession': 'P22303', 'component_descriptio...",SINGLE PROTEIN,9606
2,[],Torpedo californica,Acetylcholinesterase,16.0,False,CHEMBL4780,"[{'accession': 'P04058', 'component_descriptio...",SINGLE PROTEIN,7787
3,[],Mus musculus,Acetylcholinesterase,16.0,False,CHEMBL3198,"[{'accession': 'P21836', 'component_descriptio...",SINGLE PROTEIN,10090
4,[],Rattus norvegicus,Acetylcholinesterase,16.0,False,CHEMBL3199,"[{'accession': 'P37136', 'component_descriptio...",SINGLE PROTEIN,10116
5,[],Electrophorus electricus,Acetylcholinesterase,16.0,False,CHEMBL4078,"[{'accession': 'O42275', 'component_descriptio...",SINGLE PROTEIN,8005
6,[],Bos taurus,Acetylcholinesterase,16.0,False,CHEMBL4768,"[{'accession': 'P23795', 'component_descriptio...",SINGLE PROTEIN,9913
7,[],Anopheles gambiae,Acetylcholinesterase,16.0,False,CHEMBL2046266,"[{'accession': 'Q869C3', 'component_descriptio...",SINGLE PROTEIN,7165
8,[],Bemisia tabaci,AChE2,16.0,False,CHEMBL2366409,"[{'accession': 'B3SST5', 'component_descriptio...",SINGLE PROTEIN,7038
9,[],Leptinotarsa decemlineata,Acetylcholinesterase,16.0,False,CHEMBL2366490,"[{'accession': 'Q27677', 'component_descriptio...",SINGLE PROTEIN,7539


**Retrieve Bioactivity data for selected target**


In [6]:
# =============================================================================
# CELL 14 - CODE: Select target
# =============================================================================
# Select Human Acetylcholinesterase (CHEMBL220)
selected_target = 'CHEMBL220'
print(f"Selected Target: {selected_target}")


Selected Target: CHEMBL220


**Now retrieve only bioactivity data for target: Acetylcholinesterase (CHEMBL220) with reported IC50 values.**


In [7]:
# =============================================================================
# CELL 16 - CODE: Retrieve bioactivity data
# =============================================================================
activity = new_client.activity
results = activity.filter(target_chembl_id=selected_target).filter(standard_type="IC50")


In [9]:
# =============================================================================
# CELL 17 - CODE: Convert to DataFrame
# =============================================================================
df1 = pd.DataFrame.from_dict(results)
print(f"Total records retrieved: {len(df1)}")
df1.head(5)


Total records retrieved: 9731


,action_type,activity_comment,activity_id,activity_properties,assay_chembl_id,assay_description,assay_type,assay_variant_accession,assay_variant_mutation,bao_endpoint,...,target_organism,target_pref_name,target_tax_id,text_value,toid,type,units,uo_units,upper_value,value
0,None,None,33969,[],CHEMBL643384,Inhibitory concentration against acetylcholine...,B,None,None,BAO_0000190,...,Homo sapiens,Acetylcholinesterase,9606,None,None,IC50,uM,UO_0000065,None,0.75
1,None,None,37563,[],CHEMBL643384,Inhibitory concentration against acetylcholine...,B,None,None,BAO_0000190,...,Homo sapiens,Acetylcholinesterase,9606,None,None,IC50,uM,UO_0000065,None,0.1
2,None,None,37565,[],CHEMBL643384,Inhibitory concentration against acetylcholine...,B,None,None,BAO_0000190,...,Homo sapiens,Acetylcholinesterase,9606,None,None,IC50,uM,UO_0000065,None,50.0
3,None,None,38902,[],CHEMBL643384,Inhibitory concentration against acetylcholine...,B,None,None,BAO_0000190,...,Homo sapiens,Acetylcholinesterase,9606,None,None,IC50,uM,UO_0000065,None,0.3
4,None,None,41170,[],CHEMBL643384,Inhibitory concentration against acetylcholine...,B,None,None,BAO_0000190,...,Homo sapiens,Acetylcholinesterase,9606,None,None,IC50,uM,UO_0000065,None,0.8


In [10]:
# =============================================================================
# CELL 18 - CODE: Check standard type
# =============================================================================
df1.standard_type.unique()


array(['IC50'], dtype=object)

**Finally Save the resulting bioactivity data to a CSV file: bioactivity_raw_data.csv**


In [11]:
# =============================================================================
# CELL 20 - CODE: Save raw data
# =============================================================================
df1.to_csv('bioactivity_raw_data.csv', index=False)
print("Raw data saved!")


Raw data saved!


In [12]:
# =============================================================================
# CELL 22 - CODE: Copy to Drive
# =============================================================================
! cp bioactivity_raw_data.csv "/content/gdrive/My Drive/Colab Notebooks/data"


**Now copy "bioactivity_raw_data.csv" file to Google Drive, in folder "data"**


In [13]:
# =============================================================================
# CELL 23 - CODE: Verify file
# =============================================================================
! ls -l "/content/gdrive/My Drive/Colab Notebooks/data"


total 5657
-rw------- 1 root root 5792231 Jan 27 15:00 bioactivity_raw_data.csv


In [14]:
# =============================================================================
# CELL 24 - CODE: Preview file
# =============================================================================
! head bioactivity_raw_data.csv


action_type,activity_comment,activity_id,activity_properties,assay_chembl_id,assay_description,assay_type,assay_variant_accession,assay_variant_mutation,bao_endpoint,bao_format,bao_label,canonical_smiles,data_validity_comment,data_validity_description,document_chembl_id,document_journal,document_year,ligand_efficiency,molecule_chembl_id,molecule_pref_name,parent_molecule_chembl_id,pchembl_value,potential_duplicate,qudt_units,record_id,relation,src_id,standard_flag,standard_relation,standard_text_value,standard_type,standard_units,standard_upper_value,standard_value,target_chembl_id,target_organism,target_pref_name,target_tax_id,text_value,toid,type,units,uo_units,upper_value,value
,,33969,[],CHEMBL643384,Inhibitory concentration against acetylcholinesterase,B,,,BAO_0000190,BAO_0000357,single protein format,CCOc1nn(-c2cccc(OCc3ccccc3)c2)c(=O)o1,,,CHEMBL1148382,J Med Chem,2004.0,"{'bei': '19.61', 'le': '0.36', 'lle': '3.32', 'sei': '9.21'}",CHEMBL133897,,CHEMBL133897,6.12,0,http://www.op

# **Step 2: Data Preprocessing**
**Preprocess bioactivity data (IC50) for the selected Acetylcholinesterase target.**


**Inspect Missing Values**


In [15]:
# =============================================================================
# CELL 27 - CODE: Check missing values
# =============================================================================
print("Missing values in key columns:")
print(f"standard_type: {df1['standard_type'].isna().sum()}")
print(f"standard_value: {df1['standard_value'].isna().sum()}")
print(f"canonical_smiles: {df1['canonical_smiles'].isna().sum()}")


Missing values in key columns:
standard_type: 0
standard_value: 1286
canonical_smiles: 39


**Filter Rows with Valid Bioactivity Values**


In [16]:
# =============================================================================
# CELL 29 - CODE: Filter valid values
# =============================================================================
df2 = df1[df1["standard_value"].notna()]
print(f"Records after removing missing values: {len(df2)}")
df2.head()


Records after removing missing values: 8445


,action_type,activity_comment,activity_id,activity_properties,assay_chembl_id,assay_description,assay_type,assay_variant_accession,assay_variant_mutation,bao_endpoint,...,target_organism,target_pref_name,target_tax_id,text_value,toid,type,units,uo_units,upper_value,value
0,None,None,33969,[],CHEMBL643384,Inhibitory concentration against acetylcholine...,B,None,None,BAO_0000190,...,Homo sapiens,Acetylcholinesterase,9606,None,None,IC50,uM,UO_0000065,None,0.75
1,None,None,37563,[],CHEMBL643384,Inhibitory concentration against acetylcholine...,B,None,None,BAO_0000190,...,Homo sapiens,Acetylcholinesterase,9606,None,None,IC50,uM,UO_0000065,None,0.1
2,None,None,37565,[],CHEMBL643384,Inhibitory concentration against acetylcholine...,B,None,None,BAO_0000190,...,Homo sapiens,Acetylcholinesterase,9606,None,None,IC50,uM,UO_0000065,None,50.0
3,None,None,38902,[],CHEMBL643384,Inhibitory concentration against acetylcholine...,B,None,None,BAO_0000190,...,Homo sapiens,Acetylcholinesterase,9606,None,None,IC50,uM,UO_0000065,None,0.3
4,None,None,41170,[],CHEMBL643384,Inhibitory concentration against acetylcholine...,B,None,None,BAO_0000190,...,Homo sapiens,Acetylcholinesterase,9606,None,None,IC50,uM,UO_0000065,None,0.8


**Standardize Units - Keep only nM (nanomolar)**


In [17]:
# =============================================================================
# CELL 31 - CODE: Check and filter units
# =============================================================================
print("Unit distribution:")
print(df2['standard_units'].value_counts())

# Keep only nM units
df2 = df2[df2['standard_units'] == 'nM']
print(f"\nRecords after keeping only nM: {len(df2)}")


Unit distribution:
standard_units
nM             8374
ug.mL-1          63
10'5pM            3
10'6pM            2
10^-4microM       2
10'3pM            1
Name: count, dtype: int64

Records after keeping only nM: 8374


**Assign Bioactivity Classes**
Define active, intermediate, and inactive classes based on IC50 values.

- **Active:** IC50 ≤ 1000 nM
- **Intermediate:** 1000 < IC50 < 10000 nM  
- **Inactive:** IC50 ≥ 10000 nM


In [18]:
# =============================================================================
# CELL 33 - CODE: Assign bioactivity classes
# =============================================================================
bioactivity_class = []
for value in df2.standard_value:
    value = float(value)
    if value >= 10000:
        bioactivity_class.append("inactive")
    elif value <= 1000:
        bioactivity_class.append("active")
    else:
        bioactivity_class.append("intermediate")

print(f"Total compounds classified: {len(bioactivity_class)}")


Total compounds classified: 8374


**Extract Relevant Columns**


In [19]:
# =============================================================================
# CELL 35 - CODE: Extract columns
# =============================================================================
molecule_ids = df2.molecule_chembl_id.tolist()
canonical_smiles = df2.canonical_smiles.tolist()
standard_values = df2.standard_value.tolist()


In [20]:
# =============================================================================
# CELL 36 - CODE: Create data tuple
# =============================================================================
data = list(zip(
    molecule_ids,
    canonical_smiles,
    standard_values,
    bioactivity_class,
))


**Create Preprocessed Bioactivity Dataset**


In [21]:
# =============================================================================
# CELL 38 - CODE: Create final dataframe
# =============================================================================
df3 = pd.DataFrame(
    data,
    columns=[
        "molecule_chembl_id",
        "canonical_smiles",
        "standard_value",
        "bioactivity_class",
    ]
)
df3.head()


,molecule_chembl_id,canonical_smiles,standard_value,bioactivity_class
0,CHEMBL133897,CCOc1nn(-c2cccc(OCc3ccccc3)c2)c(=O)o1,750.0,active
1,CHEMBL336398,O=C(N1CCCCC1)n1nc(-c2ccc(Cl)cc2)nc1SCC1CC1,100.0,active
2,CHEMBL131588,CN(C(=O)n1nc(-c2ccc(Cl)cc2)nc1SCC(F)(F)F)c1ccccc1,50000.0,inactive
3,CHEMBL130628,O=C(N1CCCCC1)n1nc(-c2ccc(Cl)cc2)nc1SCC(F)(F)F,300.0,active
4,CHEMBL130478,CSc1nc(-c2ccc(OC(F)(F)F)cc2)nn1C(=O)N(C)C,800.0,active


**Remove Compounds without Valid SMILES**
Drop rows with NaN, empty, or None SMILES values.


In [22]:
# =============================================================================
# CELL 40 - CODE: Clean SMILES
# =============================================================================
print(f"Records before SMILES cleaning: {len(df3)}")

df3 = df3.dropna(subset=["canonical_smiles"])
df3 = df3[df3["canonical_smiles"].str.lower() != "none"]
df3 = df3[df3["canonical_smiles"].str.strip() != ""]

print(f"Records after SMILES cleaning: {len(df3)}")
df3.head()


Records before SMILES cleaning: 8374
Records after SMILES cleaning: 8372


,molecule_chembl_id,canonical_smiles,standard_value,bioactivity_class
0,CHEMBL133897,CCOc1nn(-c2cccc(OCc3ccccc3)c2)c(=O)o1,750.0,active
1,CHEMBL336398,O=C(N1CCCCC1)n1nc(-c2ccc(Cl)cc2)nc1SCC1CC1,100.0,active
2,CHEMBL131588,CN(C(=O)n1nc(-c2ccc(Cl)cc2)nc1SCC(F)(F)F)c1ccccc1,50000.0,inactive
3,CHEMBL130628,O=C(N1CCCCC1)n1nc(-c2ccc(Cl)cc2)nc1SCC(F)(F)F,300.0,active
4,CHEMBL130478,CSc1nc(-c2ccc(OC(F)(F)F)cc2)nn1C(=O)N(C)C,800.0,active


**Remove Duplicate Molecules**


In [23]:
# =============================================================================
# CELL 42 - CODE: Remove duplicates
# =============================================================================
print(f"Records before removing duplicates: {len(df3)}")
df3 = df3.drop_duplicates(subset=['canonical_smiles'])
print(f"Records after removing duplicates: {len(df3)}")


Records before removing duplicates: 8372
Records after removing duplicates: 6843


**Dataset Summary**


In [24]:
# =============================================================================
# CELL 44 - CODE: Summary statistics
# =============================================================================
print("=" * 50)
print("FINAL DATASET SUMMARY")
print("=" * 50)
print(f"Target: Acetylcholinesterase (CHEMBL220)")
print(f"Total curated records: {len(df3)}")
print(f"\nBioactivity class distribution:")
print(df3['bioactivity_class'].value_counts())


FINAL DATASET SUMMARY
Target: Acetylcholinesterase (CHEMBL220)
Total curated records: 6843

Bioactivity class distribution:
bioactivity_class
active          2816
inactive        2264
intermediate    1763
Name: count, dtype: int64


**Save Preprocessed Bioactivity Data**
Save the cleaned dataset to CSV and copy to Google Drive.


In [25]:
# =============================================================================
# CELL 46 - CODE: Save final data
# =============================================================================
df3.to_csv("bioactivity_preprocessed_data.csv", index=False)

!cp bioactivity_preprocessed_data.csv "/content/gdrive/My Drive/Colab Notebooks/data"
!ls "/content/gdrive/My Drive/Colab Notebooks/data"


bioactivity_preprocessed_data.csv  bioactivity_raw_data.csv


## **End of Part 1: Data Collection and Curation**

### Summary
| Step | Description |
|------|-------------|
| 1 | Searched ChEMBL for Acetylcholinesterase |
| 2 | Retrieved IC50 bioactivity data |
| 3 | Removed missing values |
| 4 | Standardized units to nM |
| 5 | Assigned bioactivity classes |
| 6 | Removed invalid SMILES |
| 7 | Removed duplicates |
| 8 | Saved curated dataset |

**Files saved:**
- `bioactivity_raw_data.csv` - Raw data from ChEMBL
- `bioactivity_preprocessed_data.csv` - Curated data ready for QSAR


| Metric                                | Value                           |
| ------------------------------------- | ------------------------------- |
| Initial Records Retrieved             | 9731                            |
| Bioactivity Type                      | IC50                            |
| Units                                 | nM (nanomolar)                  |
| Final Curated Records                 | 6843                            |
| Active Compounds (IC50 ≤ 1000 nM)     | 2816                            |
| Intermediate (1000 < IC50 < 10000 nM) | 1763                            |
| Inactive Compounds (IC50 ≥ 10000 nM)  | 2264                            |